Drop down description list with automatic types, per project.

nedSaf opened the pull request:
#569

nedSaf commented:
## Update
-  Added a multifield to the projects in order to have pre-set descriptions.
- Added a callback to get the pre-set descriptions.
- Each project that has this field populated will have this pre-set description select-list.

![edit_project_unfpa_global__hq____live_gizra_productivity](https://user-images.githubusercontent.com/7369740/40712248-e5169af2-6405-11e8-8115-b02a90340a61.png)

## Next
- Add the JS logic to check for this field on each project selected.

## TB
**1.5/5h**

nedSaf commented:
## Update
- Updated the callback item
- Added the logic for updating the select lists with the types.

## Next
- Clean up
- Updating the logic for submitting the tracking item

## TB
**_2.5/5h_**

nedSaf commented:
@RachelBaram @bricel Here's the flow now:
![pro-preselect-logic](https://user-images.githubusercontent.com/7369740/40719717-0d8a09c6-641d-11e8-8183-cfd3cdb9449f.gif)

nedSaf commented on productivity/modules/custom/productivity_tracking/plugins/content_types/tracking_form/tracking_form.inc:
```diff
@@ -50,6 +50,7 @@ function productivity_tracking_tracking_form_content_type_render($subtype, $conf
     'qa' => 'QA',
     'training' => 'Training',
   ];
+  $variables['all_types'] = list_allowed_values(field_info_field('field_issue_type'));
```
@bricel I got the "issue type" values from the `field_issue_type` for the disabled select list here because I need all the values to be present.

But I kept the hardcoded values (The array directly above this) for normal tracking because I don't want to make the list too long, do you think it's ok? Or should I handle the todo here? (`TODO: Set this to be dynamic and per project.`)

bricel replied:
> do you think it's ok

Yes, this is fine, most probably it won't apply on other projects.

RachelBaram commented:
@nedSaf Just double checking that in the backend the admins (AM) can still change the Type.

nedSaf commented:
@RachelBaram As you can see from the first screenshot, any admin can go to a project and add any pre-set description that they want as many as they want.

RachelBaram commented:
@nedSaf I am talking about a specific tracking item.  Example: the developer chose a description that autofilled the type as 'Senior QA'.  There is a special situation and Noam wants to keep the description but assign the type as 'QA'.  Will it be possible to do?
![edit_track_log_item_reivew_the_pr___live_gizra_productivity](https://user-images.githubusercontent.com/14312543/40766409-0ad71dea-64b8-11e8-923f-94f7065ad993.png)

nedSaf commented:
@RachelBaram Yes, I'll deploy it soon on a special environment (not the live one) and walk you thought it.

nedSaf commented:
Back on this:

## Next
- Clean up
- Updating the logic for submitting the tracking item
- Checking the status on view.

nedSaf commented:
## Update
- Added the save logic
- Updated the view logic (disabling fields for the pre-set description tracking)

## Next
- There's an issue while viewing multiple pre-set trackings of not displaying data of the older one :/
-  Clean up.
- Deploy to Multi-Dev (On Sunday).

nedSaf commented:
@RachelBaram This looks like it's ready, I'll deploy it soon to a multi-dev for deploying.

@bricel Do you have time for review?

nedSaf commented:
Deployed to multi-dev: 
https://preset-desc-productivity.pantheonsite.io/

AronNovak commented on productivity/modules/custom/productivity_tracking/plugins/content_types/tracking_form/tracking-table-row.tpl.php:
```diff
@@ -4,7 +4,7 @@
   </td>
   <td width="16%">
-    <select  name="project" class="form-control disable-after-save" value="<?php print $tracking['project_nid']; ?>" <?php print $tracking['disabled']; ?>>
+    <select name="project" onchange="Drupal.behaviors.trackingTable.getPresetDescriptions(this);" class="form-control disable-after-save" value="<?php print $tracking['project_nid']; ?>" <?php print $tracking['disabled']; ?>>
```
Shouldn't we do something like::
```js
$('tracking-table-selector select').change(function() {
  Drupal.behaviors.trackingTable.getPresetDescriptions(this);
});
```

from a JS file. isn't that cleaner?

nedSaf replied:
Yeah I know, I tried this in many versions in the beginning, it didn't work, for some reason the function didn't capture the change because the row is cloned using jquery.

AronNovak replied:
:ok: , then it's fine.

AronNovak commented on productivity/modules/custom/productivity_tracking/productivity_tracking.module:
```diff
@@ -29,7 +29,6 @@ function productivity_tracking_menu() {
     'file' => 'reports/productivity_tracking.burn.report.inc',
   );
 
-
```
Basically this file is not changed at all, ideally it should not be part of the diff, as the coding standard is not enforced anyways in this project. But it does not worth another commit, i think, as these changes are quite okay :)

nedSaf replied:
👍

AronNovak commented on productivity/modules/custom/productivity_project/productivity_project.module:
```diff
@@ -808,6 +816,34 @@ function productivity_project_get_projects() {
+
+  foreach ($preset_descriptions as $preset_description) {
+    $response[] = [
+      'description' => $preset_description->field_issue_description[LANGUAGE_NONE][0]['value'],
```
![image](https://user-images.githubusercontent.com/114076/40905013-e2b53d56-67dc-11e8-8acd-e21d0cd2427f.png)
![image](https://user-images.githubusercontent.com/114076/40905038-f3fb0c8a-67dc-11e8-8af5-23fd393a0235.png)

AronNovak commented on productivity/modules/custom/productivity_project/productivity_project.module:
```diff
@@ -808,6 +816,34 @@ function productivity_project_get_projects() {
+  foreach ($preset_descriptions as $preset_description) {
+    $response[] = [
+      'description' => $preset_description->field_issue_description[LANGUAGE_NONE][0]['value'],
+      'type' => $preset_description->field_issue_type[LANGUAGE_NONE][0]['value'],
```
This does not belong to this line, but in my test, the work type remained empty after saving the entry:
![image](https://user-images.githubusercontent.com/114076/40905097-34230678-67dd-11e8-8b21-1f4206488989.png)
![image](https://user-images.githubusercontent.com/114076/40905115-428fd07e-67dd-11e8-94fc-0f559eb19c36.png)
Related node on `test`: https://preset-desc-productivity.pantheonsite.io/node/137819/edit

nedSaf replied:
@AronNovak On which user I can find the entry?

AronNovak replied:
under `admin`

nedSaf replied:
Ohh I understand, but even on `live` those details are not saved on the `Tracking` entity, they are saved in the `tracking log item` entity, you can see for example:
https://productivity.gizra.com/node/137880/edit?destination=admin/content

<img width="670" alt="edit_tracking_tracking_for_project____live_gizra_productivity" src="https://user-images.githubusercontent.com/7369740/40914506-242890c0-6801-11e8-8a3a-ef451f467a14.png">

AronNovak replied:
Ok, thanks, it seems to be fine then, i just spotted the empty field with the same name, it was a false alarm then.

nedSaf commented:
Fixed the XSS issue, do you think I can merge?

AronNovak commented:
@nedSaf yep, let's trigger Travis and MWG.
